whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/DeltaImporter.java

```diff
@@ -0,0 +1,188 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeltaImporter.class);
+
+    private final NrtmRestClient nrtmRestClient;
+
+    public static final String RECORD_SEPARATOR = "\u001E";
+
+    public DeltaImporter(final NrtmRestClient nrtmRestClient,
+                         final Nrtm4ClientRepository nrtm4ClientRepository,
+                         final Nrtm4ClientInfoRepository nrtm4ClientInfoRepository) {
+        super(nrtm4ClientInfoRepository, nrtm4ClientRepository);
+        this.nrtmRestClient = nrtmRestClient;
+    }
+
+    @Override
+    public void doImport(final String source, final UpdateNotificationFileResponse updateNotificationFile){
+        final List<UpdateNotificationFileResponse.NrtmFileLink> deltas = getNewDeltas(source, updateNotificationFile);
+
+        if (deltas.isEmpty()) {
+            LOGGER.info("No new deltas to be processed");
+            return;
+        }
+
+        deltas.forEach(delta -> {
+            final byte[] deltaFilePayload = nrtmRestClient.getDeltaFile(delta.getUrl());
+
+            if (deltaFilePayload == null || deltaFilePayload.length == 0){
+                LOGGER.error("This cannot happen. UNF has a non-existing delta");
+                return;
+            }
+
+            final String payloadHash = calculateSha256(deltaFilePayload);
+            if (!delta.getHash().equals(payloadHash)){
+                LOGGER.error("Delta hash {} doesn't match the payload {}, skipping import", delta.getHash(), payloadHash);
+                return;
+            }
+
+            processPayload(deltaFilePayload,
+                    firstRecord -> {
+                        final Metadata metadata = getMetadata(firstRecord);
+                        if (!metadata.sessionId().equals(updateNotificationFile.getSessionID())){
+                            LOGGER.error("The session {} is not the same in the UNF and snapshot {}", metadata.sessionId(), updateNotificationFile.getSessionID());
+                            truncateTables();
+                            throw new IllegalArgumentException("The session is not the same in the UNF and snapshot");
+                        }
+                        nrtm4ClientInfoRepository.saveDeltaFileVersion(source, metadata.version, metadata.sessionId());
+                    }
+                    );
+
+        });
+    }
+
+    private void processPayload(final byte[] deltaFilePayload, final Consumer<String> firstRecordProcessor) {
+        ByteBuffer buffer = ByteBuffer.wrap(deltaFilePayload);
+        InputStream inputStream = new ByteArrayInputStream(buffer.array(), buffer.position(), buffer.remaining());
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
+            String record;
+            boolean isFirstRecord = true;
+            while ((record = reader.readLine()) != null) {
+                if (isFirstRecord){
+                    firstRecordProcessor.accept(record);
```

Comment:
why firstRecordProcessor? It will be use for every subsequent record right?

Reply:
no, in the first record we have the metadata.
This is just reading the metadata...which is in the first record

Reply:
discuss via slack, I will change this and do all logic in processPayload instead of passing back the first record

Reply:
Done